### ISSUE TITLE
Will onModelChange cause all the components rendered inside the factory to be re-rendered after modifying the model?

### ISSUE THREAD

**Issue by daigang1228** [caplin/FlexLayout#456]
### Describe the bug

Will onModelChange cause all the components rendered inside the factory to be re-rendered after modifying the model?

### Steps to Reproduce the Bug or Issue

Modify the layout so that you can see the content under each tab flashing

### Expected behavior

Modifying the layout should not re-render the content of each tab

### Operating System

mac

### Browser Type?

chrome

### Browser Version

128.0.6613.120

**Comment by nealus** [comment]
The model should be set, then mutated using Actions, rather than recreated for each change in the layout.

The optional onModelChange() callback is to allow processing after the model has changed (due to an action).

**Comment by daigang1228** [comment]
> 应该设置模型，然后使用操作进行变异，而不是为布局中的每个更改重新创建。
> 
> 可选的 onModelChange() 回调用于允许在模型发生变化（由于操作）后进行处理。

But there will be no issue in v0.7.3 version

**Comment by ericsvendsen** [comment]
Can confirm. Was running v0.7.15 and everything was fine. Upgraded to v0.8.1 and my app is completely hosed with "maximum update depth exceeded" error messages. Checked the changelog for breaking changes for v0.8, and nothing at all is mentioned that points to this issue. I'm fine to stay on 0.7.15 for now, but any further guidance here would be much appreciated. Thank you!

**Comment by rob-myers** [comment]
@nealus in `flexlayout-react@0.7.15` you can recreate the Model with different JSON to e.g.
- remove tab(s)
- add tabs(s)
- provide tab(s) with different props

all in-one-go, without remounting any existing tabs.

This is a _very_ helpful feature missing from `flexlayout-react@0.8.1`.

**Comment by bvanseg** [comment]
This is blocking us from upgrading to 0.8.0+. It would be very helpful if there was an action to update just the layout JSON of the model. Without the ability to do so, layout synchronization via websocket across multiple devices becomes a challenge after the changes made in 0.8.0+, as the layout updating on one end triggers the entire model on another end to be recreated.

Alternatively, maybe decoupling the layout from the model altogether would be best? Letting the model become a dispatcher for actions, while the layout itself can be updated freely.

**Comment by paralin** [comment]
I am running into the same issue. My app deterministically generates the Model after changes are applied and passes a new Model (different by reference) to the Layout. The previous versions (0.7.x) would not re-render all the tabs in this case. The new version re-renders all the tabs (unmounting the DOM and re-mounting), which is of course terrible for performance.

Reproduction of the issue: https://github.com/paralin/flex-layout-issue-3454

When you click "change layout" it passes an identical Model with Model.fromJson(sameJsonModel) and causes a re-render (new component instance created).

**Comment by paralin** [comment]
Here was what I came up with:

```
import React, { PureComponent } from 'react'
import {
  Layout,
  Model,
  TabNode,
  ILayoutProps,
  Action,
  Rect,
} from '@aptre/flex-layout'

/**
 * Optimized Layout Demo
 * 
 * This demonstrates an efficient tab rendering approach for flex-layout:
 *
 * 1. Tab Container Strategy:
 *    - Maintains a container div before the Layout component
 *    - Holds all tab content components
 *    - Uses absolute positioning for visible tabs
 *    - Keeps hidden tabs in DOM with display: none
 *
 * 2. Tab Reference System:
 *    - TabRef components manage visibility/positioning
 *    - Updates CSS properties without re-rendering content
 *    - Handles mount/unmount/resize events
 *
 * 3. Performance Benefits:
 *    - Avoids re-rendering tab contents
 *    - Only updates CSS properties for layout changes
 *    - Maintains component state when hiding/showing tabs
 */

interface ITabRefProps {
  tab: TabNode
  layout: OptimizedLayout 
}

// TabRef manages visibility and positioning of a tab
class TabRef extends PureComponent<ITabRefProps> {
  private resizeListener = (params: { rect: Rect }) => {
    this.updatePosition(params.rect)
  }

  private visibilityListener = (params: { visible: boolean }) => {
    const { tab, layout } = this.props
    if (params.visible) {
      layout.showTab(tab.getId())
    } else {
      layout.hideTab(tab.getId())
    }
  }

  componentDidMount() {
    const { tab, layout } = this.props
    layout.showTab(tab.getId())
    this.updatePosition(tab.getRect())
    tab.setEventListener('resize', this.resizeListener)
    tab.setEventListener('visibility', this.visibilityListener)
  }

  componentWillUnmount() {
    const { tab, layout } = this.props
    layout.hideTab(tab.getId())
    tab.removeEventListener('resize')
    tab.removeEventListener('visibility')
  }

  updatePosition(rect: Rect) {
    if (!rect) return
    this.props.layout.positionTab(this.props.tab.getId(), rect)
  }

  render() {
    return null
  }
}

interface ITabProps {
  tabID: string
  visible?: boolean
  rect?: Rect
  children: React.ReactNode
}

// Tab renders the actual tab content
const Tab: React.FC<ITabProps> = ({ tabID, visible, rect, children }) => {
  return (
    <div
      role="tabpanel"
      aria-labelledby={`tab-button-${tabID}`}
      data-tab-id={tabID}
      style={{
        position: 'absolute',
        display: visible ? 'flex' : 'none',
        left: rect?.x,
        top: rect?.y,
        width: rect?.width,
        height: rect?.height,
      }}
    >
      {children}
    </div>
  )
}

interface IOptimizedLayoutProps {
  model: Model
  renderTab: (tabId: string) => React.ReactNode
  layoutProps?: Partial<ILayoutProps>
}

// OptimizedLayout demonstrates the tab optimization approach
export class OptimizedLayout extends PureComponent<IOptimizedLayoutProps> {
  // Track visible tabs and their positions
  private tabRefs: {
    [tabID: string]: {
      visible: boolean
      rect?: Rect
    }
  } = {}

  showTab = (tabID: string) => {
    this.tabRefs[tabID] = {
      ...this.tabRefs[tabID],
      visible: true,
    }
    this.forceUpdate()
  }

  hideTab = (tabID: string) => {
    this.tabRefs[tabID] = {
      ...this.tabRefs[tabID],
      visible: false,
    }
    this.forceUpdate()
  }

  positionTab = (tabID: string, rect: Rect) => {
    this.tabRefs[tabID] = {
      ...this.tabRefs[tabID],
      rect,
    }
    this.forceUpdate()
  }

  private createComponent = (node: TabNode) => {
    return <TabRef tab={node} layout={this} />
  }

  private onModelChange = (model: Model, action: Action) => {
    // Handle model changes if needed
    console.log('Model changed:', action)
  }

  render() {
    const { model, renderTab, layoutProps } = this.props

    return (
      <>
        <Layout
          model={model}
          factory={this.createComponent}
          onModelChange={this.onModelChange}
          {...layoutProps}
        />
        <>
          {Object.entries(this.tabRefs).map(([tabID, ref]) => (
            <Tab
              key={tabID}
              tabID={tabID}
              visible={ref.visible}
              rect={ref.rect}
            >
              {renderTab(tabID)}
            </Tab>
          ))}
        </>
      </>
    )
  }
}
```

It renders tabs in a separate <div> and uses position: absolute to position them in the correct space when visible. When invisible we use display: none. This optimizes away the problem of Model causing a re-render of the Layout. 

I will have a look at adding this to flex-layout as well.

**Comment by rob-myers** [comment]
@paralin any update?

I'm still stuck on `0.7.15` and it looks like I have company.

<img width="757" alt="Image" src="https://github.com/user-attachments/assets/3954d1bf-d5fe-4fa4-84d1-ef9d4ccad869" />

**Comment by nealus** [comment]
In v0.7 the layout was calculated in JS code before the render, but in v0.8 the browser does the layout (using flexbox) and a second render positions the tabs (from the positions of the tabset divs). Because of this change the first render of a new model will not know the positions of the tabs, so they will not be rendered, causing a React dismount.

A possible work around is to copy the rects from the old to the new model.

let newModel = Model.fromJson(json);
copyModelRects(oldModel, newModel); // copy tabset content rects to new model


copyModelRects(fromModel: Model, toModel: Model) {
  toModel.visitNodes((toNode)=> {
      if (toNode instanceof TabSetNode || toNode instanceof BorderNode) {
          const fromNode = fromModel.getNodeById(toNode.getId()) as TabSetNode | BorderNode;
          if (fromNode) {
              toNode.setContentRect(fromNode.getContentRect());
          }
      }
  });
}


But, the model was never expected to be used this way, mutations should be applied through actions.

**Comment by paralin** [comment]
I've been using a variant of the OptimizedLayout I posted successfully. I think generally it makes sense to have two divs, one with the layout and the other with the tabs. That way when a tab is backgrounded you can set display none and keep the dom mounted (I know the base flex layout does this already to some extent). It reminds me of how a window manager works with X11. But until now I haven't found a clean way to integrate this approach directly into flex-layout.

**Comment by paralin** [comment]
I also considered adding code to compute the set of Action to mutate one Model to match another Model and fall back to your set rects workaround if the resultant Model does not 100% match. That could also work, but the number of possible edge cases makes it hard. (This is necessary because some of us are syncing the model over the network)

**Comment by cbn-falias** [comment]
TL;DR: Using [react-reverse-portal](https://www.npmjs.com/package/react-reverse-portal) "around" the layout manager should remove the mentioned issues.
Adding this in your app, could be quite easy if everything can be mounted immediately, but if tabs are loaded more dynamically a bit more work is probably required.
A long-term benefit is, that you get a lot more control over the components lifecycle and state, so that it's basically irrelevant which layout manager you use in-between.

---

In reference to the comments of @paralin all the low-level handling of the dom with `display: none` or similar, aren't required with portals.
When using `react-reverse-portal` the components are fully detached from the layout manager and within the layout you only "point" to your component.
Before: `TabNode => Component`
After: `TabNode => OutPortal ===> InPortal => Component`

How the [reproduction](https://github.com/caplin/FlexLayout/issues/456#issuecomment-2631670478) could be solved with reverse portals:
```ts
import { InPortal, OutPortal, createHtmlPortalNode } from 'react-reverse-portal';

function App() {
const [currentModel, setCurrentModel] = useState<Model>(() => Model.fromJson(BASE_MODEL));
  // NOTE: Every cmp needs its own portalNode and isn't some kind of "container for all"!
  const portalNode = useMemo(() => createHtmlPortalNode(), []);

  const setIdenticalModel = (): void => {
    setCurrentModel(Model.fromJson(BASE_MODEL));
  };

  const factory = (_tab: TabNode): React.JSX.Element => <OutPortal node={portalNode} />;

  return (
    <>
      <Layout model={currentModel} factory={factory} />
        <InPortal node={portalNode}>
          <TestTab onChangeModel={setIdenticalModel} />
        </InPortal>
    </>
  );
}
```

### Some references & learnings:
[Why to use this approach](https://www.youtube.com/watch?v=uEnLhxL8Afs)
[react-reverse-portal](https://www.npmjs.com/package/react-reverse-portal)
[Understanding portals](https://www.youtube.com/watch?v=IgJcK0SbiLM)

**Comment by cbn-falias** [comment]
@AmyLynn87 Sorry, haven't read this yet.

> But I'm still not able to find a solution about the scrollbar that jumps to the top of the panel every time I update the Model.

I guess you have to move the "overflow handling" into the portal.
The container of the Flexlayout `<div class="flexlayout__tab_moveable" />` has a `overflow: auto` by default.
1. Remove the overflow setting from the flexlayout (or set it to `hidden`)
2. Set it at the top-most container of the reverse portal
    ```ts
    createHtmlPortalNode({
              attributes: { style: 'position: relative;height:100%;width:100%;overflow:auto;' },
    });
    ```

This is totally untested so I can't tell you if it solves your issue. 😄

**Comment by brownstein** [comment]
I'm also hitting this issue. I had a workaround that was viable for `9.0.0` but it breaks in `9.0.1` because `setMovableElement` was made private. I have included the utility function I was using to allow component re-use below:

```ts
/**
 * Update handler to convert layout JSON into a Model, with an optional second
 * parameter to pull from a previous model instance to allow component re-use.
 *
 * We had to add this after upgrading to flexlayout-react@0.9.0 because they
 * now store meaningful rect and visibility data on the model.
 */
export function flexModel(layout: IJsonRowNode, prevModel: Model | null) {
  const nextModel = Model.fromJson({
    global: {
      // Ignore this section for the purpose of this example.
      tabEnableRename: false,
      tabSetEnableTabStrip: true,
      tabBorderHeight: 1,
      tabBorderWidth: 1
    },
    layout
  });

  if (prevModel) {
    nextModel.visitNodes((nextNode) => {
      const prevNode = prevModel.getNodeById(nextNode.getId());
      if (prevNode === undefined || nextNode.getType() !== prevNode.getType()) {
        return;
      }
      nextNode.setRect(prevNode.getRect());
      switch (nextNode.getType()) {
        case "tab": {
          const nextTabNode = nextNode as TabNode;
          const prevTabNode = prevNode as TabNode;
          nextTabNode.setVisible(prevTabNode.isVisible());
          nextTabNode.setTabRect(prevTabNode.getTabRect());
          nextTabNode.setMoveableElement(prevTabNode.getMoveableElement());
          break;
        }
        case "tabset": {
          const nextTabSetNode = nextNode as TabSetNode;
          const prevTabSetNode = prevNode as TabSetNode;
          nextTabSetNode.setContentRect(prevTabSetNode.getContentRect());
          break;
        }
        default:
          break;
      }
    });
  }

  return nextModel;
}
```

I think the maintainers should either adjust their code to allow replacement layouts or make `setMovableElement` public again.

A little bit about my use case: I'm using this for a game that incorporates multiple text editors that need to be moved around the screen, and I keep a copy of the model JSON in Redux. I mutate this model JSON in some Redux actions to allow listeners to animate tabs moving around the screen when triggered by entities in my levels.

**Comment by brownstein** [comment]
I have a PR up to help address this.
